Add WebID+Delegate support for TLS and OIDC modes

smalinin opened the pull request:
Added support of WebID+Delegate for both auth modes TLS and OIDC
You must have already WebID certificate installed in you browser for use `Register`,
(if you don't have WebID client certificate, you could create it free on http://ods-qa.openlinksw.com/youid/ or etc)
when new account is created, the server write Delegator profile to `card` file.
The card file URL (ex <https://myaccount.solid.openlinksw.com:8443/profile/card#me> ) must be added as relation to your WebID profile like:
```
@prefix ns1:	<http://id.myopenlink.net/DAV/home/smalinin/YouID/Sergey_Malinin_Fb_id_myopenlink_net/profile.ttl#> .

ns1:identity_cert:identity <http://www.openlinksw.com/schemas/cert#onBehalfOf>  <https://myaccount.solid.openlinksw.com:8443/profile/card#me>  .
```
The created Delegator profile may be used in browser with YouID extension https://github.com/OpenLinkSoftware/youid  for Login to Solid-Server.

dmitrizagidulin commented:
@smalinin - Thank you for this PR! Couple of style questions, before we dive into the meat of the code.
Does this PR pass the unit tests and the StandardJS linter?
I noticed that there were several sections that don't match Standard conventions. Such as the use of snake case variable and function names (for example, `handle_register_webid`) -- we only use camel case. Similarly, I noticed semicolons in several places -- Standard is strictly no-semicolon.

smalinin commented:
I will prepared the update for this pull request. Now I have an additional fixes, will fix the style and etc.

melvincarvalho commented:
Running 

    standard --fix

Should help fix a few errors

RubenVerborgh commented:
> Does this PR pass the unit tests and the StandardJS linter?

But more importantly, why didn't Travis catch this?
Did someone switch off Travis integration, or disable it for pull requests?

Worked until 5 days ago: https://travis-ci.org/solid/node-solid-server/pull_requests

dmitrizagidulin commented:
@RubenVerborgh - I was wondering that same thing, why didn't CI catch it! Does it maybe have something to do with the PR coming in from a forked repo?

RubenVerborgh commented:
PR is a PR, should work. Maybe it will magically jump back into action after the conflict is resolved.

dmitrizagidulin commented:
Ah, good point, maybe that’s it

smalinin commented:
I updated code (all last updates have been pushed):
 - style updated for camelCase
 - remove semicolons
 - all tests are passed now ( I updated 3 tests )
 - the code used rdfstore.js was replaced with new code used rdflib.js
all last updates in https://github.com/OpenLinkSoftware/node-solid-server/commit/fd0efa71d22fe0decd8064fdf36f51b6a68b6985   

Must I create new pull request or ?

melvincarvalho commented:
@smalinin that's great, no need for new PR

but the package-lock.json has conflicts

I think this must be resolved before the automated tests kick in (ie travis)

smalinin commented:
The conflict with package-lock.json may be because I was need to install global-tunnel-ng package.
It could not work with default packages on my host, after the simple `npm install`
```
npm install -s global-tunnel-ng

```

smalinin commented:
@melvincarvalho

melvincarvalho commented:
you possibly need to check in favicon.ico

smalinin commented:
Fixed package-lock.json

melvincarvalho commented:
The tests passed, excellent!

And thanks also for adding new tests.

Looks reasonable at first blush.  Although it would be good to know a summary of what has changed.  Aside from the middleware accepting delegation, there are changes to the registration and login forms I think.

I suggest we merge this and put it live so that it can be tested.  

Any reviews from @RubenVerborgh @dmitrizagidulin @timbl @ericprud welcome!

melvincarvalho commented:
@smalinin please feel free to drop into chat : 

https://gitter.im/solid/node-solid-server

If you'd prefer to discuss changes in realtime

smalinin commented:
@melvincarvalho
If I run:

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -3,7 +3,7 @@
   <meta charset="utf-8">
   <meta name="viewport" content="width=device-width, initial-scale=1">
-  <title>{{#if name}}{{name}} &ndash; {{/if}}Solid Home</title>
+  <title>Personal Data Space (or POD) {{#if name}} of {{name}} {{/if}}</title>
```
These changes are unrelated to this PR.

kidehen replied:
So can't you just remove the index.html then?

smalinin replied:
Rollback this fix

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -16,9 +16,13 @@
+        <a id="logged-href" href="" title="" style="display:none">
+          <span class="glyphicon glyphicon-user"></span>
+        </a>
+        <span>&nbsp&nbsp</span>
```
Incorrect HTML.

smalinin replied:
Fixed

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -16,9 +16,13 @@
     <div class="container-fluid">
       <div class="navbar-header">
-        <a href="/" class="navbar-brand">{{#if name}}{{name}} &ndash; {{/if}}Solid Home</a>
+        <a href="/" class="navbar-brand">Personal Data Space (POD)</a>
```
These changes are unrelated to this PR.

smalinin replied:
Rollback this fix

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -16,9 +16,13 @@
       <div class="navbar-right">
+        <a id="logged-href" href="" title="" style="display:none">
+          <span class="glyphicon glyphicon-user"></span>
+        </a>
```
Why is this needed?

smalinin replied:
It is for show user icon as href=webId, now you don't see your webId after the Login.

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -29,12 +33,28 @@
       <p class="lead">
         <span>
-          This is a public homepage of {{#if name}}{{name}}, whose WebID is{{else}}a user with WebID{{/if}}
+          This is a public homepage of <a href="{{webId}}">{{#if name}} {{name}} {{/if}}</a>
```
Why is the WebID gone?

smalinin replied:
Rollback fix to original version

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -29,12 +33,28 @@
+  <div class="row" id="update_profile" style="display:none">
+    <div class="col-md-12">
+      <p class="lead">
+        Regarding WebID-TLS+Delegation protocol, add the following RDF statement to your Delegates WebID-Profile Document or click "Update WebID-Profile Doc" to attempt this action right now.
```
Is this fixed in the template?

smalinin replied:
Rollback fix to original version

smalinin replied:
What do you mean ? The file default-templates/new-account/index.html is template for user's index.html

melvincarvalho commented on default-templates/new-account/profile/card:
```diff
@@ -23,3 +28,25 @@
+    cert:key [
+               a cert:RSAPublicKey ;
+               cert:exponent   "65537"^^xsd:int ;
+               cert:modulus    "00"^^xsd:hexBinary
```
Just checking did you want to put "00" in here or {{tlsCertModulus}} ?

smalinin replied:
Yes, now `card` profile is create as Delegator profile(that doesn't need PublicKey), so it is something like a stub only for properly work(loading data) with browser YouID extension.

RubenVerborgh replied:
That doesn't seem proper usage of modulus then though. Is there any way around this?

smalinin replied:
Yes, it is possible avoid this, I will fix our browser YouID extension

melvincarvalho commented on default-templates/new-account/index.html:
```diff
@@ -16,9 +16,13 @@
+        <a id="logged-href" href="" title="" style="display:none">
+          <span class="glyphicon glyphicon-user"></span>
+        </a>
+        <span>&nbsp&nbsp</span>
```
possible semi colon here

smalinin replied:
Fixed

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -122,17 +151,100 @@ <h1>Apps</h1>
+        logged_href.href = session.webId;
+        logged_href.title = session.webId;
+      }
     }
```
A bit too much magic here. Why not separate Log In and Log Out buttons that we show and hide, instead of one multipurpose button?

smalinin replied:
The "magic" was added for show user icon(like anchor with href=webId)  on the left side of Login/Logout button.  The code for Login/Logout button was already in the original version.

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -122,17 +151,100 @@ <h1>Apps</h1>
   function sessionActionButton() {
     return document.getElementById('session-action')
   }
+  function sessionLogged_href() {
```
Naming is off.

smalinin replied:
Commented code was removed

RubenVerborgh commented on default-templates/new-account/index.html:
```diff
@@ -122,17 +151,100 @@ <h1>Apps</h1>
+          alert('Attr: '+response.statusText);
+        })
+        .catch(function(error) {})
+***/
```
Please remove.

smalinin commented:
I will prepare the Readme for created changes.

RubenVerborgh commented:
What are all the favicon changes about?

smalinin commented:
@RubenVerborgh 
The favicon.ico(.acl) wasn't in the server root https://localhost  before, it was only in the user home https://user.localhost

RubenVerborgh commented:
I see. We might need to consider splitting this PR in multiple ones, as it seems to address multiple concerns.

melvincarvalho commented on lib/models/account-template.js:
```diff
@@ -79,10 +79,20 @@ class AccountTemplate {
   static templateSubstitutionsFor (userAccount) {
+    let addOwlSameAs = userAccount.externalWebId &&
+           userAccount.tls_certificate_webid &&
+           userAccount.externalWebId !== userAccount.tls_certificate_webid
```
Is this part correct.  Should it be camel case?

smalinin replied:
Yes, it is a bug, I missed this rename, the fix has been commited

RubenVerborgh commented on default-views/auth/login-tls.hbs:
```diff
@@ -1,36 +1,92 @@
+            try {
+              const url = new URL(text);
+              window.location.href = url.href;
+            } catch(err) {
```
What can go wrong here?

RubenVerborgh commented on lib/models/authenticator.js:
```diff
@@ -192,11 +196,40 @@ class TlsAuthenticator extends Authenticator {
+        .then(exists => {
+          if (!exists) {
+            debug(`Account ${this.username} is not existed`)
+            throw new Error(`Account ${this.username} is not existed`)
```
* does not exist

RubenVerborgh commented on lib/models/authenticator.js:
```diff
@@ -293,10 +329,54 @@ class TlsAuthenticator extends Authenticator {
+    var delegator
+
+    if (me.headers) {
+      delegator = me.headers['on-behalf-of'] || me.headers['On-Behalf-Of'] || null
```
Should only need to be one case (lowercase).

RubenVerborgh commented on lib/webid_util.js:
```diff
@@ -0,0 +1,266 @@
+        }
+      }
+    }
+    return {name}
```
We should avoid this pyramid code, that's hard to maintain. We could for instance put all those predicates in an array and do a reduce.

smalinin replied:
Fixed

RubenVerborgh commented on lib/create-server.js:
```diff
@@ -28,6 +29,10 @@ function createServer (argv, app) {
   }
 
+  if (argv.httpProxy) {
+    globalTunnel.initialize(argv.httpProxy)
```
We should document this config option.

smalinin commented:
About updates in PR (I hope, I didn't forgot anything)

- updated WebID-TLS and WebID-OIDC auth for support WebID+Delegate-TLS and WebID+Delegate-OIDC

- index.html was updated:
    -  show user icon(like anchor with href=webId) on the left side of Login/Logout button.
    - added button "Update WebID-Profile doc" for add next Delegator relation
```
<TLS-WebID-URI>   <http://www.openlinksw.com/schemas/cert#onBehalfOf>   <https://user.localhost:8444/profile/card#me> .
```

- Auth OIDC
    - Register:
        - added loading TLS WebID or DelegatorWebID as default value to field "External WebID"
        - the "card" file is updated for using it as Delegator profile
        - if ExternalWebID was a Delegator, the next relation is added to all .acl files
       `<#owner> acl:agent  <ExternalWebID> .`
    - Login:
        - Username filed could not be empty now for both UID/PWD Login and WebID+Delegate-TLS Login. It is needed for propely redirect to userHome,  if ExternalWebID is used for Login
       
- Auth TLS
    - Register was not implemented before
       - now it create "card" profile, that could be used as Delegator profile
       - the UI field "Name" is loaded by default with value from TLS WebId(Certificate.CN)  or Delegator (name from profile data)
       - the UI field "WebID" is loaded by default with TLS WebID or Delegator WebID
       - if ExternalWebID was a Delegator, the next relation is added to all .acl files
       `<#owner> acl:agent  <ExternalWebID> .`
   - Login was not implemented before
        - Username filed could not be empty now for both UID/PWD Login and WebID+Delegate-TLS Login. It is needed for propely redirect to userHome, if ExternalWebID is used for Login

melvincarvalho commented:
@smalinin changelog looks great, thanks! 

you might want to add a line about fixing the favicon too

RubenVerborgh commented:
@smalinin Do we have a README though with a scenario?

melvincarvalho commented:
@smalinin good progress!  

Do you have this branch (4.1) running on a live server?

It would be good to try and get it running so we can test live, and in parallel explain how it can be tested.

kidehen commented:
@melvincarvalho, 

Our live servers are:
1. https://solid.openlinksw.com:8443 -- WebID-TLS+Delegation IdP
1. https://solid.openlinksw.com:8444 -- WebID-OIDC + WebID-TLS+Delegation Idp

Right now, there are live instances from this latest release at:

1. https://kidehen3.solid.openlinksw.com:8443
1. https://kidehen5.solid.openlinksw.com:8444

smalinin commented:
I am sorry, but I was need to add back to removed fix, that was mark as outdated,
because in authMode = OIDC, the Login with Certificate(WebID+Delegate TLS) doesn't work .
The Server return back redirect command and WebID in HTTP Header **User**, when server authMode== OIDC, the HTTP Header **User** is removed by browser from HTTP Response Headers. I don't have idea, why it is happened, because the same things works properly when server authMode == TLS.  
Therefore I added fix back to sources. 
Now server doesn't sent back HTTP 302 redirect, it returns back the redirect URL as String, the client code receives this string and execute redirect.

dmitrizagidulin commented:
> in authMode = OIDC, the Login with Certificate(WebID+Delegate TLS) doesn't work

Could part of the issue be - when authMode = OIDC, the Express app doesn't request the TLS cert from the connection? Or are you passing in the cert in the cert header somehow?

smalinin commented on default-templates/new-account/public/.acl:
```diff
@@ -6,6 +6,9 @@
 <#owner>
     a acl:Authorization;
     acl:agent <{{webId}}>;
+{{#if addOwlSameAs}}
```
If you Register new account with externalWebID from Delegator, 
the code from **{{#if addOwlSameAs}}** is added. 
If you Register new account with externalWebID from TLS Certificate, it isn't added.
p.s.
It is done, because I could not add owl:sameAs relation for externalWebID(that is the Delegator also). **owl:sameAs** isn't supported by rdflib.js

smalinin replied:
I added the next relation to **card** profile (if externalWebID is from Delegator)
```
:me  owl:sameAs <${externalWebID}> .
```

smalinin commented:
@dmitrizagidulin 
The browser ask me the TLS certificate, when I click on button **[With Certificate (WebID+Delegate-TLS)]** in Login page.

melvincarvalho commented:
@smalinin Thanks for this awesome contribution and for responding quickly to feedback!

Spoke to @RubenVerborgh today and we agreed to merge this into the v4.1.0 branch and test it further.

I think it's possible to add more changes.

Our hope is to then merge 4.1 into master and upgrade the version.

melvincarvalho commented:
Is it possible to leave this issue open, as it's not yet been merged into master?

Or shall we perhaps start a fresh issue?

melvincarvalho commented:
Ah It's a PR.  I think we should associate an issue with it?

dmitrizagidulin commented on default-templates/new-account/settings/.acl:
```diff
@@ -6,6 +6,9 @@
 
     acl:agent
         <{{webId}}>;
+{{#if addOwlSameAs}}
```
I still think this section does not need to be added to *all* the .acls

dmitrizagidulin commented on lib/webid_util.js:
```diff
@@ -0,0 +1,255 @@
+ * A class wrapping utility routines for WebID handling.
+ * @constructor
+ */
+function WebIdUtils () {
```
Would it make sense to refactor this to use ES6 `class` syntax, to match the rest of the code base?
